libraries/lib-components/EffectInterface.h

```diff
@@ -408,6 +408,15 @@ class COMPONENTS_API EffectInstance
    // Suggest a block size, but the return is the size that was really set:
    virtual size_t SetBlockSize(size_t maxBlockSize) = 0;
 
+   //! How many input buffers to allocate at once
+   /*!
+    If the effect ALWAYS processes channels independently, this can return 1
+    */
+   virtual unsigned GetAudioInCount() const = 0;
+
+   //! How many output buffers to allocate at once
+   virtual unsigned GetAudioOutCount() const = 0;
+
```

Comment:
The commit that moves this virtual function to the Instance class is surprising.  I thought the number of channels in and out is a property of a plug-in independent of its instances.  What problem led to this change?

Reply:
In VST3 this property belongs to the instance. Also the number of IO channels is configurable

Reply:
To understand that, I need more explanation, or an example of a plug-in.

I'm not comfortable with this change.  I need to reexamine all the uses of these virtual functions to be sure they make sense.

Alternatively, if the number of channels in and out can be described as a function of EffectSettings, independently of an Instance state, that might make more sense.

I can make modifications of the branch so that it can build without this commit.  That is, I have not discovered where you need to call these functions as part of the implementation of other overridden virtual functions in the instance.

Reply:
I'm reconsidering this, and I think you do make a change for the better.  I have noticed some confusion about the meaning of these virtual functions.

But ultimately, `EffectInstance::ProcessBlock` can be reached (for destructive effects) or `EffectInstance::RealtimeProcess` and those functions are not given a count of channels.  Instead the instance assumes that it "knows" the right number of buffers in and out.

I think I want to cherry-pick this commit, and do some other things relating to these virtual functions, and propse a PR to be merged first, and then the rest of your work can rebase onto that.

Reply:
However there should be some kind of contract expaining when it is permissible to call these virtual functions on an instance.

Where does the change of VST3 instance state happen, that determines the result of `CountChannels`?  It is not at construction time of the Instance, because no information is given to `MakeInstance`.  It must be one of the later functions that pass Settings, yes?

Reply:
Sorry, I mean that it's possible in VST3 to change channel arrangement dynamically, but we don't(yet) support this feature.
https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/Change+History/3.0.0/Multiple+Dynamic+IO.html?highlight=channel#audio-bus-channel-configuration